### ISSUE TITLE
noUnusedLocals checks don't work with private modifier

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#15775]
**TypeScript Version:**  master

**Code**

```ts
class C {
  constructor(private parent: void) {
    parent;
  }
}
```

**Expected behavior:**

pass

**Actual behavior:**

```
$ node built/local/tsc.js --noEmit --noUnusedLocals index.ts
index.ts(2,23): error TS6138: Property 'parent' is declared but never used.
```

**Comment by mhegazy** [comment]
`this.parent` is never used.

**Comment by falsandtru** [comment]
`constructor(public/protected parent: void)` and `constructor(parent: void)` work correctly. Only `constructor(private parent: void)` throws an error. At least you should arrange the behaviors.

**Comment by mhegazy** [comment]
`constructor(public/protected parent: void)`  =>  public/ protected properties can be used outside the class, so they are not flagged as unused. the same applies to exported declarations in a module.

`constructor(parent: void)`  => `parent` is a parameter, you use it inside the constructor, no error here..


` constructor(private parent: void)` => is equivalent to:
```ts
class C {
  private parent;
  constructor(parent: void) {
  }
}
```

if you never use `this.parent` why did you declare it as a property.

**Comment by falsandtru** [comment]
> parent is a parameter

`node built/local/tsc.js --noEmit --noUnusedParameters --noUnusedLocals index.ts` doesn't throw an error with `constructor(parent: void)`.  But throws an error if I add private modifier. It is odd.

**Comment by mhegazy** [comment]
`private parent` declares two things....
1. a parameter to this function, which you have used ..
2. a property on the class, which you have not used. hence the error: ` Property 'parent' ....`

I suggest you read more about parameter properties in http://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

**Comment by falsandtru** [comment]
> two things

I understood now, thanks. But it is not confusing and it can be resolved. I think almost entire users will expect resolving this issue.

**Comment by falsandtru** [comment]
In the following code, `noUnusedParameters` checks resolved by `noUnusedLocals`. But the converse is not.

```ts
class C {
  constructor(private parent: void) {
    this.parent;
  }
}
```

Compiler should resolve these two checks bidirectionally.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
